File: app/src/main/java/com/hover/stax/pushNotification/FirebaseMessageReceiver.java
package com.hover.stax.pushNotification;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.widget.RemoteViews;

import androidx.annotation.NonNull;
import androidx.core.app.NotificationCompat;

import com.google.firebase.messaging.FirebaseMessaging;
import com.google.firebase.messaging.FirebaseMessagingService;
import com.google.firebase.messaging.RemoteMessage;
import com.hover.stax.R;
import com.hover.stax.SplashScreenActivity;
import com.hover.stax.utils.Constants;
import com.hover.stax.utils.DateUtils;

import java.util.Date;
import java.util.Map;

public class FirebaseMessageReceiver extends FirebaseMessagingService {
	@Override
	public void onNewToken(@NonNull String s) {
		super.onNewToken(s);
	}

	@Override
	public void
	onMessageReceived(RemoteMessage remoteMessage) {
		Map<String, String> data = remoteMessage.getData();
		String redirect = data.get("redirect");
		if (remoteMessage.getNotification() != null) {
			showNotification(remoteMessage.getNotification().getTitle(), remoteMessage.getNotification().getBody(), redirect);
		}
	}

	private void showNotification(String title, String message, String redirect) {
		String channel_id = String.valueOf(DateUtils.now());

		Intent intent = new Intent(this, SplashScreenActivity.class);
		intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
		intent.putExtra(Constants.FROM_FCM, title);
		intent.putExtra(Constants.FRAGMENT_DIRECT, redirect);
		PendingIntent pendingIntent = PendingIntent.getActivity(this, 0, intent, PendingIntent.FLAG_ONE_SHOT);

		NotificationCompat.Builder builder = new NotificationCompat.Builder(getApplicationContext(), channel_id)
													 .setSmallIcon(R.drawable.ic_stax)
													 .setAutoCancel(false)
													 .setBadgeIconType(NotificationCompat.BADGE_ICON_LARGE)
													 .setPriority(2)
													 .setTicker(title)
													 .setOngoing(false)
													 .setUsesChronometer(false)
													 .setVibrate(new long[]{1000, 1000, 1000, 1000, 1000})
													 .setOnlyAlertOnce(false)
													 .setContentTitle(title)
													 .setContentText(message)
													 .setContentIntent(pendingIntent);

		NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
			NotificationChannel notificationChannel = new NotificationChannel(channel_id, getString(R.string.app_name), NotificationManager.IMPORTANCE_HIGH);
			if (notificationManager != null)
				notificationManager.createNotificationChannel(notificationChannel);
		}
		if (notificationManager != null) {
			int randId = (int) ((new Date().getTime() / 1000L) % Integer.MAX_VALUE);
			notificationManager.notify(randId, builder.build());
		}
	}
}

File: app/src/main/java/com/hover/stax/notifications/FirebaseMessageReceiver.kt
package com.hover.stax.notifications

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.os.bundleOf
import androidx.navigation.NavDeepLinkBuilder
import com.google.firebase.messaging.FirebaseMessagingService
import com.google.firebase.messaging.RemoteMessage
import com.hover.stax.R
import com.hover.stax.RoutingActivity
import com.hover.stax.financialTips.FinancialTipsFragment
import com.hover.stax.utils.Constants
import com.hover.stax.utils.DateUtils
import timber.log.Timber
import kotlin.random.Random

class FirebaseMessageReceiver : FirebaseMessagingService() {

    override fun onNewToken(newToken: String) {
        super.onNewToken(newToken)
        // update token on backend if used for notifications
    }

    override fun onMessageReceived(message: RemoteMessage) {
        val data = message.data
        val redirect = data["redirect"]

        if (message.notification != null) {
            Timber.e("Showing notification message")
            showNotification(message.notification!!.title!!, message.notification!!.body!!, redirect)
        } else {
            Timber.e("Showing data message")
            showNotification(data["title"]!!, data["body"]!!, redirect)
        }
    }

    private fun showNotification(title: String, message: String, redirect: String?) {
        val channelId: String = DateUtils.now().toString()

        val pendingIntent = getPendingIntent(title, redirect)

        val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        createNotificationChannel(channelId, notificationManager)

        val builder = NotificationCompat.Builder(applicationContext, channelId).apply {
            setSmallIcon(R.drawable.ic_stax)
            setAutoCancel(true)
            setBadgeIconType(NotificationCompat.BADGE_ICON_LARGE)
            priority = 2
            setTicker(title)
            setOngoing(false)
            setUsesChronometer(false)
            setVibrate(longArrayOf(1000, 1000, 1000, 1000, 1000))
            setOnlyAlertOnce(true)
            setContentTitle(title)
            setContentText(message)
            setContentIntent(pendingIntent)
        }

        notificationManager.notify(Random.nextInt(), builder.build())
    }

    private fun createNotificationChannel(channelId: String, notificationManager: NotificationManager) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(channelId, getString(R.string.app_name), NotificationManager.IMPORTANCE_HIGH)
            notificationManager.createNotificationChannel(channel)
        }
    }

    private fun getPendingIntent(title: String, redirect: String?): PendingIntent {
        Timber.e("Redirect: $redirect")

        return if (redirect != null && redirect.contains(getString(R.string.deeplink_financial_tips))) {
            val tipId = Uri.parse(redirect).getQueryParameter("id")

            NavDeepLinkBuilder(this)
                .setGraph(R.navigation.home_navigation)
                .setDestination(R.id.wellnessFragment)
                .setArguments(bundleOf(FinancialTipsFragment.TIP_ID to tipId))
                .createPendingIntent()
        } else {
            val intent = Intent(this, RoutingActivity::class.java).apply {
                addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
                putExtra(Constants.FROM_FCM, title)
                putExtra(Constants.FRAGMENT_DIRECT, redirect)
            }
            PendingIntent.getActivity(this, 0, intent, PendingIntent.FLAG_ONE_SHOT)
        }
    }

}
